core/src/main/java/org/geysermc/geyser/session/GeyserSession.java

```diff
@@ -879,6 +884,14 @@ public boolean onMicrosoftLoginComplete(PendingMicrosoftAuthentication.Authentic
      * After getting whatever credentials needed, we attempt to join the Java server.
      */
     private void connectDownstream() {
+        SessionLoginEvent loginEvent = new SessionLoginEvent(this, remoteServer);
+        GeyserImpl.getInstance().eventBus().fire(loginEvent);
+        if (loginEvent.isCancelled()) {
+            disconnect("Login event cancelled");
```

Comment:
For clarity, I'm fine if we have a method in SessionLoginEvent to set the disconnect reason in `setCancelled` (i.e. `setCancelled(true, "Kicked")`, I just don't want a getter for it since it makes deprecating it harder in the future. Would also use one of the generic bedrock disconnect messages in the case that none is specified.